docs/namespace_config_spec.md

```diff
@@ -0,0 +1,319 @@
+
+The Ory Permission Language is a syntactical subset of TypeScript. Along with type
+definitions for the syntax elements of the language (such as `Namespace` or
+`Context`), users can get context help from their IDE while writing the
+configuration.
+
+## Notation
+
+The syntax is specified using Extended Backus-Naur Form (EBNF):
+
+```ebnf
+Production  = production_name "=" [ Expression ] "." .
+Expression  = Alternative { "|" Alternative } .
+Alternative = Term { Term } .
+Term        = production_name | token [ "…" token ] | Group | Option | Repetition .
+Group       = "(" Expression ")" .
+Option      = "[" Expression "]" .
+Repetition  = "{" Expression "}" .
+```
+
+Productions are expressions constructed from terms and the following operators,
+in increasing precedence:
+
+```ebnf
+|   alternation
+()  grouping
+[]  option (0 or 1 times)
+{}  repetition (0 to n times)
+```
+
+Lower-case production names are used to identify lexical tokens. Non-terminals
+are in CamelCase. Lexical tokens are enclosed in double quotes `""` or single quotes `''`.
+
+The form `a … b` represents the set of characters from a through b as
+alternatives. The horizontal ellipsis `…` is also used elsewhere in the spec to
+informally denote various enumerations or code snippets that are not further
+specified.
+
+## Configuraton text representation
+
+The configuration is encoded in UTF-8.
+
+## Lexical elements
+
+### Comments
+
+1. Line comments start with the character sequence // and stop at the end of the
+   line.
+2. General comments start with the character sequence /* and stop with the first
+   subsequent character sequence */.
+3. Documentation comments start with the character sequence /** and stop with
+   the first subsequent character sequence */.
+
+### Identifiers
+
+Identifiers name program entities such as variables and types. An identifier is
+a sequence of one or more letters and digits. The first character in an
+identifier must be a letter.
+
+```ebnf
+identifier = letter { letter | digit } .
+digit      = "0" … "9" .
+letter     = "A" … "Z" | "a" … "z" | "_" .
+```
+
+### String literals
+
+String literals represent string constants as sequences of characters.
+
+```ebnf
+string_lit    = single_quoted | double_quoted .
+single_quoted = "'" { letter | digit } "'" .
+double_quoted = '"' { letter | digit } '"' .
+```
+
+### Keywords
+
+The configuration language has the following keywords:
+
+* `class`
+* `implements`
+* `metadata`
+* `related`
+* `permits`
+* `this`
+* `ctx`
+* `id`
+
+### Builtin Types
+
+The following types are built in:
+
+* `Context`
+* `Namespace`
+* `Namespace[]`
+* `boolean`
+* `string`
+
+In TypeScript, they would be defined as follows:
+
+```typescript
+type Context = { subject: never }
+
+interface Namespace {
+    metadata: { id: string }
+    related?: { [relation: string]: Namespace[] }
+    permits?: { [method: string]: (ctx: Context) => boolean }
+}
+
+interface Array<Namespace> {
```

Comment:
The type definition is not part of the configuration language, it is just part of the tooling. We never have to parse the type definitions. Therefore, we can still use the `Array<...>` syntax here, since this is part of TypeScript.